File: 01_Acey_Ducey/pascal/simple/aceyducey.pas
program aceyducey;

{$IFDEF FPC}
{$mode objfpc}{$H+}
{$ENDIF}

uses
 Crt;

var
  Stash: Integer;
  CardA: Integer;
  CardB: Integer;
  CardC: Integer;
  Bet: Integer;

procedure PrintGreeting;
begin
  WriteLN(' ':26, 'ACEY DUCEY CARD GAME');
  WriteLN(' ':15, 'CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY');
  WriteLN;
  WriteLN;
  WriteLN('ACEY-DUCEY IS PLAYED IN THE FOLLOWING MANNER ');
  WriteLN('THE DEALER (COMPUTER) DEALS TWO CARDS FACE UP');
  WriteLN('YOU HAVE AN OPTION TO BET OR NOT BET DEPENDING');
  WriteLN('ON WHETHER OR NOT YOU FEEL THE CARD WILL HAVE');
  WriteLN('A VALUE BETWEEN THE FIRST TWO.');
  WriteLN('IF YOU DO NOT WANT TO BET, INPUT A 0');
  WriteLN;
end;

procedure PrintBalance;
begin
  WriteLN('YOU NOW HAVE ', Stash,' DOLLARS.');
  WriteLN;
end;

procedure PrintCard(const ACard: Integer);
begin
  if ACard < 11 then
  begin
    Write(ACard);
  end;
  if ACard = 11 then
  begin
    Write('JACK');
  end;
  if ACard = 12 then
  begin
    Write('QUEEN');
  end;
  if ACard = 13 then
  begin
    Write('KING');
  end;
  if ACard = 14 then
  begin
    Write('ACE');
  end;
end;

procedure DrawDealerCards;
var
  tmp: Integer;
begin
  Write('HERE ARE YOUR NEXT TWO CARDS: ');
  repeat
    CardA:= Random(14) + 2;
  until (CardA >= 2) and (CardA <= 14);
  repeat
    CardB:= Random(14) + 2;
  until (CardB >= 2) and (CardB <= 14) and (CardA <> CardB);
  if CardA > CardB then
  begin
    tmp:= CardB;
    CardB:= CardA;
    CardA:= tmp;
  end;
  PrintCard(CardA);
  Write(' ');
  PrintCard(CardB);
  WriteLN;
  WriteLN;
end;

procedure DrawPlayerCard;
begin
  repeat
    CardC:= Random(14) + 2;
  until (CardC >= 2) and (CardC <= 14);
  PrintCard(CardC);
  WriteLN;
  WriteLN;
end;

function GetBet: Integer;
begin
  Result:= 0;
  repeat
    Write('WHAT IS YOUR BET: ');
    ReadLN(Result);
    if Result > Stash then
    begin
      WriteLn('SORRY, MY FRIEND, BUT YOU BET TOO MUCH.');
      WriteLn('YOU HAVE ONLY ', Stash,' DOLLARS TO BET.');
    end;
  until (Result >=0) and (Result <= Stash);
end;

function TryAgain: Boolean;
var
  answer: String;
begin
  Result:= False;
  Write('TRY AGAIN (YES OR NO)');
  ReadLn(answer);
  Result:= (LowerCase(answer)='yes') or (LowerCase(answer)='y');
end;

begin
  Randomize;
  ClrScr;
  PrintGreeting;
  repeat
    Stash:= 100;
    repeat
      PrintBalance;
      DrawDealerCards;
      Bet:= GetBet;
      if Bet = 0 then
      begin
        WriteLN('CHICKEN!!');
        continue;
      end;
      DrawPlayerCard;
      if (CardC > CardA) and (CardC < CardB) then
      begin
        WriteLN('YOU WIN!!!');
        Inc(Stash, Bet)
      end
      else
      begin
        WriteLN('SORRY, YOU LOSE');
        Dec(Stash, Bet)
      end;
    until Stash = 0;
    WriteLN('SORRY, FRIEND, BUT YOU BLEW YOUR WAD.');
    WriteLN;
  until not TryAgain;
  WriteLN('O.K., HOPE YOU HAD FUN!');
end.

File: 00_Alternate_Languages/01_Acey_Ducey/nim/aceyducey.nim
import std/[random,strutils]

var
  bet, cardA, cardB, cardC, stash: int
  retry: bool = true

randomize() # Seed the random number generator

proc printGreeting() =
  echo spaces(26),"ACEY DUCEY CARD GAME"
  echo spaces(15),"CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY"
  echo """

ACEY-DUCEY IS PLAYED IN THE FOLLOWING MANNER
THE DEALER (COMPUTER) DEALS TWO CARDS FACE UP
YOU HAVE AN OPTION TO BET OR NOT BET DEPENDING
ON WHETHER OR NOT YOU FEEL THE CARD WILL HAVE
A VALUE BETWEEN THE FIRST TWO.
IF YOU DO NOT WANT TO BET, INPUT A 0

"""

proc printBalance() =
  echo "YOU NOW HAVE ", stash," DOLLARS."
  echo ""

proc printCard(aCard: int) =
  case aCard:
  of 11: echo "=== JACK ==="
  of 12: echo "=== QUEEN ==="
  of 13: echo "=== KING ==="
  of 14: echo "=== ACE ==="
  else: echo "=== ", aCard, " ==="

proc drawDealerCards() =
  echo "HERE ARE YOUR NEXT TWO CARDS: "
  cardA = rand(2..14)
  cardB = cardA # Copy cardA, so we can test cardB to be different
  while cardB == cardA:
    cardB = rand(2..14)
  if cardA > cardB: # Make sure cardA is the smaller card
    swap cardA, cardB
  echo ""
  printCard cardA
  echo ""
  printCard cardB
  echo ""

proc drawPlayerCard() =
  cardC = rand 2..14
  printCard cardC
  echo ""

proc getBet(): int =
  result = stash + 1 #ensure we enter the loop
  while (result < 0) or (result > stash):
    echo "WHAT IS YOUR BET: "
    result = readLine(stdin).parseInt()
    if result > stash:
      echo "SORRY, MY FRIEND, BUT YOU BET TOO MUCH."
      echo "YOU HAVE ONLY ", stash, " DOLLARS TO BET."
    if result == 0:
      echo "CHICKEN!!"

proc tryAgain(): bool =
  echo "TRY AGAIN (YES OR NO)"
  var answer = readLine(stdin).normalize()
  result = (answer == "y") or (answer == "yes")

printGreeting()
while retry:
  stash = 100
  while stash > 0:
    printBalance()
    drawDealerCards()
    bet = getBet()
    echo ""
    drawPlayerCard()
    if (cardC >= cardA) and (cardC <= cardB):
      echo "YOU WIN!!!"
      stash += bet
    else:
      echo "SORRY, YOU LOSE"
      stash -= bet
  echo "SORRY, FRIEND, BUT YOU BLEW YOUR WAD."
  echo ""
  retry = tryAgain()
echo "O.K., HOPE YOU HAD FUN!"
